### PR TITLE
An initial quantized matmul implementation

### PR DESCRIPTION
**Pull request by angeloskath**
This adds a first version for a quantized `Q(matrix) @ vector` product kernel and `matrix @ Q(matrix)^T` product kernel.

The initial API is

```
quantized_matmul(x: array, w: array, scales: array, biases: array, width: int, groups: int)
```

`w` should be `uint32` and together with `scales` and `biases` is the quantized representation of a matrix. `w` is expected to be transposed for now and `x` should be either `1xD` or `MxD` for `M % 32 == 0`.

#### Initial speed results

When microbenchmarking matrix-vector product we have good initial speedup. The following numbers are in seconds on my M2 air.

```
                     |  FP32 | FP16 | B8_G128 | B8_G64 | B4_G128 | B4_G64 | B2_G128 | B2_G64 
---------------------------------------------------------------------------------------------
1x14336 @ 14336x4096 |  2.68 | 1.38 |    0.93 |   0.95 |    0.52 |   0.54 |    0.53 |   0.57
1x4096  @ 4096x14336 |  2.65 | 1.33 |    0.78 |   0.78 |    0.51 |   0.52 |    0.53 |   0.58
```

P.S.: I will rebase on main when round is merged so ignore those changes for now.

### CODE REVIEW

File: python/src/ops.cpp [hunk 1]
```diff
@@ -2951,4 +2951,33 @@ void init_ops(py::module_& m) {
...
+      "groups"_a = 128,
+      "width"_a = 4,
+      py::kw_only(),
+      "stream"_a = none,
+      R"pbdoc(
+        quantized_matmul(x: array, w: array, scales: array, biases: array, /, groups: int = 128, width: int = 4, *, stream: Union[None, Stream, Device] = None) -> array
+
+        Perform the matrix multiplication of with the quantized matrix w. The
+        quantization uses one floating point scale and bias per ``groups`` of
+        elements. Each element in w takes ``width`` bits and is packed in q
```
Comment (awni) [inline python/src/ops.cpp:2971]: could you double back tick on ``w`` as well.

File: python/src/ops.cpp [hunk 2]
```diff
@@ -2951,4 +2951,33 @@ void init_ops(py::module_& m) {
...
+      "biases"_a,
+      py::pos_only(),
+      "groups"_a = 128,
+      "width"_a = 4,
+      py::kw_only(),
+      "stream"_a = none,
+      R"pbdoc(
+        quantized_matmul(x: array, w: array, scales: array, biases: array, /, groups: int = 128, width: int = 4, *, stream: Union[None, Stream, Device] = None) -> array
+
+        Perform the matrix multiplication of with the quantized matrix w. The
```
Comment (awni) [inline python/src/ops.cpp:2969]: typo: multiplication with

File: python/src/ops.cpp (hunk 1 above)
Comment (awni) [inline python/src/ops.cpp:2971]: What is `q` referring to here?

File: python/src/ops.cpp [hunk 3]
```diff
@@ -2951,4 +2951,33 @@ void init_ops(py::module_& m) {
...
+        unsigned 32 bit integer.
+
+        Args:
+          x (array): Input array
+          w (array): Quantized matrix packed in unsigned integers
+          scales (array): The scales to use per ``groups`` elements of ``w``
+          biases (array): The biases to use per ``groups`` elements of ``w``
+          groups (int): The size of the group in ``w`` that shares a scale and
+                        bias. (default: 128)
+          width (int): The bitwidth of the elements in ``w``. (default: 4)
```
Comment (awni) [inline python/src/ops.cpp:2981]: Add `Returns:` here as well.

**Review by awni — APPROVED** [review]
This is beautifully done! Just wow.

Just left a few comments on the docstring.

File: python/src/ops.cpp (hunk 1 above)
Comment (angeloskath) [reply python/src/ops.cpp:2971]: Hehe nothing it is referring to the fact that `q` is close to `a` in the keyboard :-D . Either way it should be "in an unsigned 32-bit integer" I guess.

**Comment by angeloskath** [comment]
@jagrit06 since this is a pretty big one, let me know if you see anything easily fixable in the kernels. Otherwise I will fix Awni's comments and merge it as a first version.

**Review by jagrit06 — APPROVED** [review]
Love it! 
No quick changes come to mind - just a few small optimizations that we can work on later down the line. We should merge it in for now!

Great work again

**Comment by angeloskath** [comment]
Fixed the comments and also fixed `qmm` to allow for `x` to be of arbitrary batch shape.
E.g. if x is `(8, 7, K)` should now result in `(8, 7, N)`.

**Pull request merged**